core/src/main/java/com/linecorp/armeria/internal/common/LeakTracingRequestContextStorage.java

```diff
@@ -0,0 +1,110 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import static com.linecorp.armeria.internal.common.RequestContextUtil.newIllegalContextPushingException;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.RequestContextStorage;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.Sampler;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.concurrent.FastThreadLocal;
+
+/**
+ * A {@link RequestContextStorage} which keeps track of {@link RequestContext}s, reporting pushed thread
+ * information if a {@link RequestContext} is leaked.
+ */
+@UnstableApi
+final class LeakTracingRequestContextStorage implements RequestContextStorage {
+
+    private final RequestContextStorage delegate;
+    private final FastThreadLocal<PendingRequestContextStackTrace> pendingRequestCtx;
+    private final Sampler<? super RequestContext> sampler;
+
+    /**
+     * Creates a new instance.
+     * @param delegate the underlying {@link RequestContextStorage} that stores {@link RequestContext}
+     * @param sampler the {@link Sampler} that determines whether to retain the stacktrace of the context leaks
+     */
+    LeakTracingRequestContextStorage(RequestContextStorage delegate,
+                                            Sampler<? super RequestContext> sampler) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        pendingRequestCtx = new FastThreadLocal<>();
+        this.sampler = requireNonNull(sampler, "sampler");
+    }
+
+    @Nullable
+    @Override
+    public <T extends RequestContext> T push(RequestContext toPush) {
+        requireNonNull(toPush, "toPush");
+
+        final RequestContext prevContext = delegate.currentOrNull();
+        if (prevContext != null) {
+            if (prevContext == toPush) {
+                // Re-entrance
+            } else if (toPush instanceof ServiceRequestContext &&
+                       prevContext.root() == toPush) {
+                // The delegate has the ServiceRequestContext whose root() is toPush
+            } else if (toPush instanceof ClientRequestContext &&
+                       prevContext.root() == toPush.root()) {
+                // The delegate has the ClientRequestContext whose root() is the same as toPush.root()
+            } else {
+                throw newIllegalContextPushingException(prevContext, toPush, pendingRequestCtx.get());
+            }
+        }
+
+        if (sampler.isSampled(toPush)) {
+            pendingRequestCtx.set(new PendingRequestContextStackTrace(toPush, true));
```

Comment:
When `push()` method is invoked in a Kotlin Coroutine with `ArmeriaRequestCoroutineContext`, 
only meaningless stack traces are left. So even if we provide the stack traces of the
previous context that was leaked, it may not help to detect the leak point of a `CoroutineContext`.

I figured out that [kotlinx-coroutines-debug](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-debug/index.html) can capture 
the stack traces of coroutines from which we can know the actual method. 
For example:
```kt
override fun updateThreadContext(context: CoroutineContext): SafeCloseable {
    val coroutineInfos = DebugProbes.dumpCoroutinesInfo()
    try {
        return requestContext.push()
    } catch (cause: IllegalStateException) {
        for (coroutineInfo in coroutineInfos) {
            if (coroutineInfo.context[Key] != null) {
                val stackTraceElements = coroutineInfo.lastObservedStackTrace()
                cause.setStackTrace(stackTraceElements.toTypedArray())
            }
        }
        throw cause
    }
}
```
This trick could be used to create `PendingRequestContextStackTrace` as well.

Before implementing it, we need to move `ArmeriaRequestCoroutineContext` to
the core module with optional Kotlin dependency. Because Armeria has two 
`ArmeriaRequestCoroutineContext` implementations in `armeria-kotlin` and `armeria-grpc`.

As this PR has a lot of changes already, we may support it in the f/u PR.